### PR TITLE
CASSANDRA-16842 Allow CommitLogSegmentReader to optionally skip sync marker CRC checks

### PR DESCRIPTION
**Pull request by maedhroz**
patch by Caleb Rackliffe; reviewed by Josh McKenzie for CASSANDRA-16842

Co-authored-by: Jordan West <jordanrw@gmail.com>
Co-authored-by: Caleb Rackliffe <calebrackliffe@gmail.com>
Co-authored-by: Marcus Eriksson <marcuse@apache.org>

### CODE REVIEW

File: src/java/org/apache/cassandra/db/commitlog/CommitLogSegmentReader.java [hunk 1]
```diff
@@ -151,8 +165,22 @@ private int readSyncMarker(CommitLogDescriptor descriptor, int offset, RandomAcc
         updateChecksumInt(crc, (int) reader.getPosition());
         final int end = reader.readInt();
         long filecrc = reader.readInt() & 0xffffffffL;
+
         if (crc.getValue() != filecrc)
         {
+            // The next marker position and CRC value are not writen atomically, so it is possible for the latter to
```
Comment (jmckenzie-dev) [inline src/java/org/apache/cassandra/db/commitlog/CommitLogSegmentReader.java]: nit: written

File: src/java/org/apache/cassandra/db/commitlog/CommitLogSegmentReader.java [hunk 2]
```diff
@@ -151,8 +165,22 @@ private int readSyncMarker(CommitLogDescriptor descriptor, int offset, RandomAcc
...
         {
+            // The next marker position and CRC value are not writen atomically, so it is possible for the latter to 
+            // still be zero after the former has been finalized, even though the mutations that follow it are valid.
+            // When there is no compression or encryption enabled, we can ignore a sync marker CRC mismatch and defer 
+            // to the per-mutation CRCs, which may be preferable to preventing startup altogether.
+            if (allowSkipSyncMarkerCrc
+                && descriptor.compression == null && !descriptor.getEncryptionContext().isEnabled()
+                && filecrc == 0 && end != 0)
+            {
+                logger.warn("Skipping sync marker CRC check at position {} (end={}, calculated crc={}) of commit log {}...",
```
Comment (jmckenzie-dev) [inline src/java/org/apache/cassandra/db/commitlog/CommitLogSegmentReader.java]: Consider adding something to the log about "Using per mutation CRC checks to ensure correctness instead"

File: test/unit/org/apache/cassandra/db/commitlog/CommitLogTest.java [hunk 3]
```diff
@@ -764,6 +782,75 @@ public void replaySimple() throws IOException
...
+        List<String> activeSegments = CommitLog.instance.getActiveSegmentNames();
+        assertFalse(activeSegments.isEmpty());
+
+        File directory = new File(CommitLog.instance.segmentManager.storageDirectory);
+        File firstActiveFile = Objects.requireNonNull(directory.listFiles((file, name) -> activeSegments.contains(name)))[0];
+        zeroFirstSyncMarkerCRC(firstActiveFile);
+
+        CommitLogSegmentReader.setAllowSkipSyncMarkerCrc(true);
+        
+
```
Comment (jmckenzie-dev) [inline test/unit/org/apache/cassandra/db/commitlog/CommitLogTest.java]: nit: extra line

**Review by jmckenzie-dev — APPROVED** [review]
A few formatting nits but otherwise LGTM

**Comment by maedhroz** [comment]
Committed as https://github.com/apache/cassandra/commit/f9b7c1e6984f5b81aae1e3a2191d4e9599db15ae

**Pull request closed without merging**